modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rho_loss_downsampling.py

```diff
@@ -0,0 +1,69 @@
+from typing import Optional, Union
+
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.irreducible_loss_producer import IrreducibleLossProducer
+
+
+class RemoteRHOLossDownsampling(AbstractRemoteDownsamplingStrategy):
+    """
+    Method adapted from Prioritized Training on Points that are Learnable, Worth Learning, and Not Yet Learnt
+    (Sören Mindermann+, 2022). https://arxiv.org/abs/2206.07137
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        modyn_config: dict,
+        per_sample_loss: torch.nn.modules.loss,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, modyn_config, device)
+        self.per_sample_loss_fct = per_sample_loss
+        self.irreducible_loss_producer = IrreducibleLossProducer(
+            per_sample_loss,
+            modyn_config,
+            params_from_selector["rho_pipeline_id"],
+            params_from_selector["il_model_id"],
+            device,
+        )
+        self.rho_loss: torch.Tensor = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def init_downsampler(self) -> None:
+        self.index_sampleid_map: list[int] = []
+        self.rho_loss = torch.tensor([])
+        self.number_of_points_seen = 0
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_input: Union[dict[str, torch.Tensor], torch.Tensor],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        training_loss = self.per_sample_loss_fct(forward_output, target).detach()
+        self.index_sampleid_map += sample_ids
+        irreducible_loss = self.irreducible_loss_producer.get_irreducible_loss(sample_ids, forward_input, target)
+        self.rho_loss = torch.cat([self.rho_loss, training_loss - irreducible_loss]).to(training_loss.dtype)
+        self.number_of_points_seen += forward_output.shape[0]
+
+    def select_points(self) -> tuple[list[int], torch.Tensor]:
+        target_size = max(int(self.downsampling_ratio * self.number_of_points_seen / 100), 1)
+        # find the indices of maximal "target_size" elements in the list of rho_loss
+        selected_indices = torch.topk(self.rho_loss, target_size).indices
```

Comment:
In a follow up PR, we could also try, instead of top k selection as in the paper, to sample based on the loss as weights, so similar to the LossDownsampler where the higher the RHO-LOSS here, the higher the probability of being sampled, but no top k. We could add a config option for this. Let's also add a config option at some point whether we want to finetune the previous IL model. Let's create two issues for this so we don't forget about it :)

Reply:
https://github.com/eth-easl/modyn/issues/504